WalletWasabi.Backend/Controllers/WabiSabi/IdempotencyRequestCache.cs

```diff
@@ -0,0 +1,107 @@
+using Microsoft.Extensions.Caching.Memory;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.WabiSabi.Backend.Models;
+using WalletWasabi.WabiSabi.Crypto;
+
+namespace WalletWasabi.Backend.Controllers.WabiSabi
+{
+	public class IdempotencyRequestCache
+	{
+		public delegate Task<TResponse> ProcessRequestDelegateAsync<TRequest, TResponse>(TRequest sender, CancellationToken cancellationToken);
+
+		/// <summary>Timeout specifying how long a request response can stay in memory.</summary>
+		private static TimeSpan CacheTimeout { get; } = TimeSpan.FromMinutes(5);
+
+		private static TimeSpan RequestTimeout { get; } = TimeSpan.FromMinutes(1);
+
+		public IdempotencyRequestCache(IMemoryCache cache)
+		{
+			ResponseCache = cache;
+		}
+
+		/// <summary>Guards <see cref="ResponseCache"/>.</summary>
+		/// <remarks>Unfortunately, <see cref="CacheExtensions.GetOrCreate{TItem}(IMemoryCache, object, Func{ICacheEntry, TItem})"/> is not atomic.</remarks>
+		/// <seealso href="https://github.com/dotnet/runtime/issues/36499"/>
+		private object ResponseCacheLock { get; } = new();
+
+		/// <remarks>Guarded by <see cref="ResponseCacheLock"/>.</remarks>
+		private IMemoryCache ResponseCache { get; }
+
+		/// <typeparam name="TRequest">
+		/// <see langword="record"/>s are preferred as <see cref="object.GetHashCode"/>
+		/// and <see cref="object.Equals(object?)"/> are generated for <see langword="record"/> types automatically.
+		/// </typeparam>
+		/// <typeparam name="TResponse">Type associated with <typeparamref name="TRequest"/>. The correspondence should be 1:1 mapping.</typeparam>
+		public async Task<TResponse> GetCachedResponseAsync<TRequest, TResponse>(TRequest request, ProcessRequestDelegateAsync<TRequest, TResponse> action, CancellationToken cancellationToken = default)
+			where TRequest : notnull
+			where TResponse : class
+		{
+			bool callAction = false;
+			TaskCompletionSource<TResponse> responseTcs;
```

Comment:
I wouldn't use a `TaskCompletionSource` but the original `Task` instead.

Reply:
Could you elaborate on that?

Reply:
I think Lucas means: instead of setting `callAction = true`, the task could just be started there and put into `ResponseCache` as `Task<TResponse>`, which does seem simpler to me

Reply:
Hm, I don't get it. I use `callAction` to know when new `TaskCompletionSource` was created and thus I need to call the callback. I think it's actually easier to undestand than having something nullable for example.

Reply:
We don't need a `TaskCompletionSource` because we already have a `Task`, otherwise we have to await for that `TaskCompletionSource`'s task and set its result  just because we use the `TaskCompletionSource`'s task instead of the original task. We can just wait for the original task.

Reply:
I put a new instance of `TaskCompletionSource` to `ResponseCache` to express "somebody is working on working on given request already". So it does not seem that simple as you say. But again, you are free to make a code suggestion, if I don't understand correctly.